### PR TITLE
cmd/snap-update-ns: apply content mounts before layouts

### PR DESCRIPTION
**Pull request by bboozzoo**
Tweak how mounts coming form content and layouts are ordered. Specifically make
sure that entries coming from content mounts are applied before ones from the
layouts (x-snapd.origin=layout). This ensures that that once snap-update-ns starts setting up layout, all the content (from content snaps) is already in place and can be referenced in the layout

### CODE REVIEW

File: cmd/snap-update-ns/change.go [hunk 1]
```diff
@@ -667,23 +694,27 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
...
 
+		logger.Debugf("adding mimics and their entries")
+
 		for _, mimicDir := range allMimicCreationDirs {
 			// make sure to sort the entries for each mimic dir in a consistent
 			// order
 			entries := entriesForMimicDir[mimicDir]
 			sort.Sort(byOriginAndMagicDir(entries))
+			logger.Debugf("mimic creation dirs")
+			for _, en := range allMimicCreationDirs {
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:705]: this looks wrong here, was it meant to be before the for loop? or to be entries?

File: cmd/snap-update-ns/change.go [hunk 2]
```diff
@@ -499,9 +499,19 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
 		desired[i].Dir = filepath.Clean(desired[i].Dir)
 	}
 
+	logger.Debugf("current profile:\n%v", current)
+	logger.Debugf("desired profile")
+	for _, en := range desired {
+		logger.Debugf("- %v", en)
+	}
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:506]: we should have a small helper for this, to make the new debug code a bit less distracting

**Review by pedronis — COMMENTED** [review]
this still looks reasonable at the simplest possible adjustment to me, some small detail comment about the new debug logging

File: cmd/snap-update-ns/change.go [hunk 3]
```diff
@@ -694,27 +691,26 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
 
 		sort.Strings(allMimicCreationDirs)
 
-		logger.Debugf("adding mimics and their entries")
+		logger.Debugf("all mimics:")
+		for _, en := range allMimicCreationDirs {
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:695]: en is is miminDir here, no? needs a different name

File: cmd/snap-update-ns/change.go [hunk 4]
```diff
@@ -694,27 +691,26 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
...
 				addDesiredEntry(entry)
 			}
 		}
 	}
 
 	sort.Sort(byOriginAndMagicDir(newIndependentDesiredEntries))
-
-	for _, entry := range append(newIndependentDesiredEntries, newDesiredEntries...) {
+	allEntries := append(newIndependentDesiredEntries, newDesiredEntries...)
+	dumpMountEntries(allEntries, "final mount entries")
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:712]: "mount entries ordered as they will be applied"

**Review by pedronis — APPROVED** [review]
lgtm, thanks, small remarks about the logging

File: cmd/snap-update-ns/change.go [hunk 5]
```diff
@@ -601,24 +611,36 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
...
 	addDesiredEntry := func(entry osutil.MountEntry) {
 		if !addedDesiredEntries[entry.Dir] {
+			logger.Debugf("adding entry: %s", entry)
 			newDesiredEntries = append(newDesiredEntries, entry)
 			addedDesiredEntries[entry.Dir] = true
 		}
 	}
+	addIndependentDesiredEntry := func(entry osutil.MountEntry) {
+		if !addedDesiredEntries[entry.Dir] {
+			logger.Debugf("adding independent entry: %s", entry)
```
Comment (mardy) [inline cmd/snap-update-ns/change.go:627]: nitpick^2: if you also printed `entry.XSnapdOrigin()` in this line, you could get rid of the line
```
logger.Debugf("- overname entry: %v", entry)
```
below.

**Review by mardy — APPROVED** [review]
This is fantastic, and also (relatively) easy to read. :-) Thanks!

File: cmd/snap-update-ns/change.go (hunk 5 above)
Comment (mardy) [reply cmd/snap-update-ns/change.go:627]: Actually, this is printed already, as it's part of the mount options; so you could just remove the line below, I don't think it adds anything.

**Review by mvo5 — APPROVED** [review]
Thank you

**Pull request merged**